src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -15,103 +15,129 @@ namespace System.Net.Http
-            // It will check if propagation is on (because parent Activity exists or there is a listener) or off (forcibly disabled)
-            // This code won't be called unless consumer unsubscribes from DiagnosticListener right after the check.
-            // So some requests happening right after subscription starts might not be instrumented. Similarly,
-            // when consumer unsubscribes, extra requests might be instrumented
+            bool currentActivitySet = Activity.Current is not null;
+            bool diagnosticListenerEnabled = s_diagnosticListener.IsEnabled();
 
-            if (request == null)
+            if (!currentActivitySet && !diagnosticListenerEnabled)
             {
-                throw new ArgumentNullException(nameof(request), SR.net_http_handler_norequest);
+                return false;
             }
 
-            Activity? activity = null;
-            DiagnosticListener diagnosticListener = Settings.s_diagnosticListener;
+            bool diagnosticListenerActivityEnabled = diagnosticListenerEnabled && s_diagnosticListener.IsEnabled(ActivityName, request);
 
-            // if there is no listener, but propagation is enabled (with previous IsEnabled() check)
-            // do not write any events just start/stop Activity and propagate Ids
-            if (!diagnosticListener.IsEnabled())
+            if (diagnosticListenerActivityEnabled || currentActivitySet)
             {
-                activity = new Activity(DiagnosticsHandlerLoggingStrings.ActivityName);
-                activity.Start();
-                InjectHeaders(activity, request);
+                activity = s_activitySource.CreateActivity(ActivityName, ActivityKind.Client);
 
-                try
+                if (activity is null && (diagnosticListenerActivityEnabled || !s_activitySource.HasListeners()))
                 {
-                    return async ?
-                        await base.SendAsync(request, cancellationToken).ConfigureAwait(false) :
-                        base.Send(request, cancellationToken);
-                }
-                finally
-                {
-                    activity.Stop();
+                    // ActivitySource decided not to create an Activity
+                    // If that is because there were no listeners, manually create an Activity to maintain the behaviour before ActivitySource was introduced
+                    // If a diagnostics listener is enabled for the Activity, always create one
+                    activity = new Activity(ActivityName);
                 }
             }
 
-            Guid loggingRequestId = Guid.Empty;
+            if (activity is null)
+            {
+                // There is no Activity, but we may still want to use the instrumented SendAsyncCore if diagnostic listeners are interested in other events
+                return diagnosticListenerEnabled;
```

Comment:
If it is possible to reach this point where currentActivitySet == true I would propose we should still be injecting headers. However if the logic changes above based on the other comments this condition will never happen and there is nothing to worry about.